### PR TITLE
Reschedulable tasks

### PR DESCRIPTION
**Pull request by bakulf**
This is the algorithm I was discussing with you. It seems to work fine. Not for 2.10.

### CODE REVIEW

File: src/tasks/group/taskgroup.cpp [hunk 1]
```diff
@@ -51,9 +51,35 @@ void TaskGroup::cancel() {
   }
 }
 
-bool TaskGroup::deletable() const {
+Task::DeletePolicy TaskGroup::deletePolicy() const {
+  // The policy for task group is:
+  // - if there is at least 1 non-deletable task, the group is non-deletable
+  // - if the tasks are all reschedulable, the group is reschedulable
+  // - if there are no reschedulable tasks, the group is deletable
+  // - fallback: the task is not deletable
```
Comment (brizental) [inline src/tasks/group/taskgroup.cpp:59]: Hum, it is not obvious to me why you landed on this policy for groups. Would you mind expanding?

File: src/taskscheduler.cpp [hunk 2]
```diff
@@ -79,19 +79,40 @@ void TaskScheduler::taskCompleted() {
...
-  if (m_running_task && m_running_task->deletable()) {
+  if (m_running_task && m_running_task->deletePolicy() == Task::Deletable) {
     m_running_task->cancel();
     m_running_task->deleteLater();
     m_running_task->disconnect();
     m_running_task = nullptr;
   }
+
+  if (!reschedulableTasks.isEmpty()) {
+    QTimer::singleShot(500, this, [this, reschedulableTasks]() {
```
Comment (brizental) [inline src/taskscheduler.cpp:112]: Why wait 500 seconds pushing the task to the end of the line? Also, what if we have multiple reschedulable tasks in a row? One single run of this timer would reschedule all of them, but there would be multiple timers running, no? Wouldn't it make more sense to schedule a timer per task and ditch the `reschedulabeTasks` list?

File: tests/unit/testtasks.cpp [hunk 3]
```diff
@@ -104,8 +104,58 @@ void TestTasks::function() {
   QVERIFY(completed);
 }
 
-void TestTasks::removeDevice() {
```
Comment (brizental) [inline tests/unit/testtasks.cpp:107]: Would you mind filing a ticket as a reminder that we need to write removeDevice task tests?

File: tests/unit/testtasks.cpp [hunk 4]
```diff
@@ -104,8 +104,58 @@ void TestTasks::function() {
...
+
+  // The first one is t2.
+  QCOMPARE(sequence.at(0), 2);
+
+  // Then we have t4 because t3 is rescheduable.
+  QCOMPARE(sequence.at(1), 4);
+
+  // The 2 reschedulable tasks: t3 and the sentinel.
+  QCOMPARE(sequence.at(2), 3);
+  QCOMPARE(sequence.at(3), 99);
```
Comment (brizental) [inline tests/unit/testtasks.cpp:158]: I think there should also be tests for the TaskGroup delete policy.

File: src/tasks/group/taskgroup.cpp (hunk 1 above)
Comment (bakulf) [reply src/tasks/group/taskgroup.cpp:59]: My idea is that the group should not override the policy of the sub-tasks.
1. The first point seems reasonable: if there is at least 1 non-deletable task the whole group is non-deletable.
2. The second one makes sense too: if all are reschedulable, the group is reschedulable.
3. The third one seems correct: if all are deletable -> let's delete the entire group
4. the last one is the complex one. If we are here, we have some reschedulable tasks, and some deletable ones. We don't know if changing the order of the execution we can introduce bugs, so my approach is: do nothing, keep the group as it is.

File: tests/unit/testtasks.cpp (hunk 3 above)
Comment (bakulf) [reply tests/unit/testtasks.cpp:107]: https://mozilla-hub.atlassian.net/browse/VPN-2896

File: src/tasks/group/taskgroup.cpp (hunk 1 above)
Comment (brizental) [reply src/tasks/group/taskgroup.cpp:59]:
```suggestion
  // - if there is at least 1 reschedulable task, the group in non-deletable
```
Just to make it clear.

File: tests/unit/testtasks.cpp [hunk 5]
```diff
@@ -104,8 +105,97 @@ void TestTasks::function() {
...
+                     new TaskFunction([&]() {}, Task::Reschedulable)});
+    QCOMPARE(g->deletePolicy(), Task::Reschedulable);
+    delete g;
+  }
+
+  // Not deletable group if mixed
+  {
+    TaskGroup* g = new TaskGroup(
+        QList<Task*>{new TaskFunction([&]() {}, Task::NonDeletable),
+                     new TaskFunction([&]() {}, Task::Reschedulable)});
```
Comment (brizental) [inline tests/unit/testtasks.cpp:195]: This group falls on the same code path as the group in line 172. Maybe do Deletable, Reschedulable instead?

**Pull request merged**